plugin/storage/badger/spanstore/reader.go

```diff
@@ -0,0 +1,488 @@
+
+	"github.com/jaegertracing/jaeger/model"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+// All of this is replicated from the ES and Cassandra storage parts.. they should be refactored to a common place
+
+var (
+	// ErrServiceNameNotSet occurs when attempting to query with an empty service name
+	ErrServiceNameNotSet = errors.New("Service Name must be set")
+
+	// ErrStartTimeMinGreaterThanMax occurs when start time min is above start time max
+	ErrStartTimeMinGreaterThanMax = errors.New("Start Time Minimum is above Maximum")
+
+	// ErrDurationMinGreaterThanMax occurs when duration min is above duration max
+	ErrDurationMinGreaterThanMax = errors.New("Duration Minimum is above Maximum")
+
+	// ErrMalformedRequestObject occurs when a request object is nil
+	ErrMalformedRequestObject = errors.New("Malformed request object")
+
+	// ErrStartAndEndTimeNotSet occurs when start time and end time are not set
+	ErrStartAndEndTimeNotSet = errors.New("Start and End Time must be set")
+
+	// ErrUnableToFindTraceIDAggregation occurs when an aggregation query for TraceIDs fail.
+	ErrUnableToFindTraceIDAggregation = errors.New("Could not find aggregation of traceIDs")
+
+	// ErrNotSupported during development, don't support every option - yet
+	ErrNotSupported = errors.New("This query parameter is not supported yet")
+
+	errNoTraces = errors.New("No trace with that ID found")
+
+	defaultMaxDuration = model.DurationAsMicroseconds(time.Hour * 24)
+)
+
+const (
+	defaultNumTraces = 100
+	sizeOfTraceID    = 16
+)
+
+// TraceReader reads traces from the local badger store
+type TraceReader struct {
+	store *badger.DB
+	cache *CacheStore
+}
+
+// NewTraceReader returns a TraceReader with cache
+func NewTraceReader(db *badger.DB, c *CacheStore) *TraceReader {
+	return &TraceReader{
+		store: db,
+		cache: c,
+	}
+}
+
+// getTraces enriches TraceIDs to Traces
+func (r *TraceReader) getTraces(traceIDs []model.TraceID) ([]*model.Trace, error) {
+	// Get by PK
+	traces := make([]*model.Trace, 0, len(traceIDs))
+	prefixes := make([][]byte, 0, len(traceIDs))
+
+	for _, traceID := range traceIDs {
+		prefixes = append(prefixes, createPrimaryKeySeekPrefix(traceID))
+	}
+
+	err := r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchSize = 10 // TraceIDs are not sorted, pointless to prefetch large amount of values
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		val := []byte{}
+		for _, prefix := range prefixes {
+			spans := make([]*model.Span, 0, 4) // reduce reallocation requirements by defining some initial length
+
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				// Add value to the span store (decode from JSON / defined encoding first)
+				// These are in the correct order because of the sorted nature
+				item := it.Item()
+				val, err := item.ValueCopy(val)
+				if err != nil {
+					return err
+				}
+
+				sp := model.Span{}
+				switch item.UserMeta() & 0x0F {
```

Comment:
What is 0x0F ?

Reply:
I only want the last 4 bits of the UserMeta byte to be able to use switch for the encoding type. If there was no 0x0F, then that switch statement wouldn't match